File: src/main/java/travelator/analytics/MarketingAnalytics.java
package travelator.analytics;

import java.util.List;
import java.util.Map;
import java.util.stream.Stream;

import static java.util.stream.Collectors.groupingBy;

public class MarketingAnalytics {
    private final EventStore eventStore;

    public MarketingAnalytics(
        EventStore eventStore
    ) {
        this.eventStore = eventStore;
    }

    public double averageNumberOfEventsPerCompletedBooking(
        String timeRange
    ) {
        Stream<Map<String, Object>> eventsForSuccessfulBookings =
            eventStore
                .queryAsStream("type=CompletedBooking&timerange=" + timeRange)
                .flatMap(event -> {
                    String interactionId = (String) event.get("interactionId");
                    return eventStore.queryAsStream("interactionId=" + interactionId);
                });
        Map<String, List<Map<String, Object>>> bookingEventsByInteractionId =
            eventsForSuccessfulBookings.collect(groupingBy(
                event -> (String) event.get("interactionId"))
            );
        var averageNumberOfEventsPerCompletedBooking =
            bookingEventsByInteractionId
                .values()
                .stream()
                .mapToInt(List::size)
                .average();
        return averageNumberOfEventsPerCompletedBooking.orElse(Double.NaN);
    }
}

File: src/main/java/travelator/analytics/MarketingAnalytics.kt
package travelator.analytics

import java.util.stream.Collectors.groupingBy

class MarketingAnalytics(
    private val eventStore: EventStore
) {
    fun averageNumberOfEventsPerCompletedBooking(
        timeRange: String
    ): Double {
        val eventsForSuccessfulBookings = eventStore
            .queryAsStream("type=CompletedBooking&timerange=$timeRange")
            .flatMap { event ->
                val interactionId = event["interactionId"] as String
                eventStore.queryAsStream("interactionId=$interactionId")
            }
        val bookingEventsByInteractionId = eventsForSuccessfulBookings.collect(
            groupingBy { event -> event["interactionId"] as String }
        )
        val values = bookingEventsByInteractionId.values
        return averageBy(values) { it.size }
    }
}

private fun averageBy(
    values: Collection<MutableList<MutableMap<String, Any>>>,
    selector: (MutableList<MutableMap<String, Any>>) -> Int
): Double {
    return values.sumBy(selector) / values.size.toDouble()
}